kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt

```diff
@@ -5,52 +5,253 @@ import io.kotest.matchers.Matcher
 import io.kotest.matchers.MatcherResult
 import io.kotest.matchers.should
 
-infix fun <T : Comparable<T>, I : Iterable<T>> I.shouldHaveUpperBound(t: T): I {
-   toList().shouldHaveUpperBound(t)
+// BooleanArray left out in the interest of reducing API bloat.
+// As Boolean only has 2 values, it is more natural to use
+// "shouldContain true / shouldContain false"-type assertions
+
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun ByteArray.shouldHaveUpperBound(value: Byte): ByteArray {
+   asList() should haveUpperBound(value, "ByteArray")
+   return this
+}
+
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun ShortArray.shouldHaveUpperBound(value: Short): ShortArray {
+   asList() should haveUpperBound(value, "ShortArray")
+   return this
+}
+
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun CharArray.shouldHaveUpperBound(value: Char): CharArray {
+   asList() should haveUpperBound(value, "CharArray")
+   return this
+}
+
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun IntArray.shouldHaveUpperBound(value: Int): IntArray {
+   asList() should haveUpperBound(value, "IntArray")
+   return this
+}
+
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun LongArray.shouldHaveUpperBound(value: Long): LongArray {
+   asList() should haveUpperBound(value, "LongArray")
    return this
 }
 
-infix fun <T : Comparable<T>> Array<T>.shouldHaveUpperBound(t: T): Array<T> {
-   asList().shouldHaveUpperBound(t)
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun FloatArray.shouldHaveUpperBound(value: Float): FloatArray {
+   asList() should haveUpperBound(value, "FloatArray")
    return this
 }
 
-infix fun <T : Comparable<T>, C : Collection<T>> C.shouldHaveUpperBound(t: T): C {
-   this should haveUpperBound(t)
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun DoubleArray.shouldHaveUpperBound(value: Double): DoubleArray {
+   asList() should haveUpperBound(value, "DoubleArray")
    return this
 }
 
-fun <T : Comparable<T>, C : Collection<T>> haveUpperBound(t: T) = object : Matcher<C> {
-   override fun test(value: C): MatcherResult {
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun <T : Comparable<T>> Array<T>.shouldHaveUpperBound(value: T): Array<T> {
```

Comment:
Do we need to have this overload? I think Iterable would catch it, right?

Reply:
No, `Array` does not implement `Iterable` - `Array<MyType>` is just `MyType[]`.

Reply:
In general a collection assertion has to have these overloads to cover all "reasonable" receivers:

1) primitive arrays
2) Array<T>
3) Iterable<T> (also covers all Collections and Ranges)
3) Sequence<T> (not an Iterable, but has iterator - cheaply representable as Iterable. Big opportunity for reusing the Iterable Matcher implementation under the hood.)
4) If appropriate, Map<K, T> - if the assertion makes sense to treat Map as a collection of Entries. Otherwise, Map doesn't impement Iterable on its own.